src/rnnlm/sample_a_word.cc

```diff
@@ -0,0 +1,432 @@
+// sample_a_word.cc
+
+#include "sample_a_word.h"
+#include <iostream>
+#include <string>
+#include <iterator>
+#include <algorithm>
+#include <math.h>
+
+// Constructor for sampling the next word
+NgramModel::NgramModel(char* arpa_file, char* histories_file) {
+  vocab_size_ = 0;
+  ReadARPAModel(arpa_file);
+  ReadHistories(histories_file);
+}
+
+// Read language model from a ARPA-format file.
+void NgramModel::ReadARPAModel(char* file) {
+  std::ifstream data_input(file);
+  if (!data_input.is_open()) {
+    std::cerr << "error opening '" << file
+              << "' for reading\n";
+    exit(1);
+  }
+  std::string line;
+  int32 order;
+  int32 order_current = 0;
+  int32 word;
+  int32 iter = 0;
+  int32 while_iter = 0;
+  std::pair <float, float> probs_pair;
+  float log_prob;
+  float backoff_weight;
+  bool unigram_check = false;
+  std::cout << "Start reading ARPA-format file..." << std::endl;
+  while (getline(data_input, line)) {
+    std::istringstream is(line);
+    // get the strings splitted by single space
+    // brace-initialization with C++11
+    std::istream_iterator<std::string> begin(is), end;
+    std::vector<std::string> tokens(begin, end);
+    if (tokens.size() == 0) continue;
+    if (tokens.size() == 2 && tokens[0] == "ngram") {
+      std::string substring = tokens[1].substr(2);
+      int32 count = std::stoi(substring); // get "123456" from "1=123456"
+      counts_.push_back(count);
+      order = std::stoi(tokens[1].substr(0));
+      continue;
+    }
+    if (tokens.size() == 1 && tokens[0] == "\\1-grams:") {
+      ngram_order_ = order; // ngram_order
+      probs_.resize(ngram_order_);
+      std::cout << "Ngram order is: " << ngram_order_ << std::endl;
+    }
+    // read current order
+    if (tokens.size() == 1 && tokens[0] != "\\data\\" &&
+        tokens[0] != "\\end\\") {
+      order_current = std::stoi(tokens[0].substr(1,1));
+      continue; // get the order info and skip processing this line
+    }
+    // read vocab and initialize probs of unigrams
+    if (order_current == 1) {
+      std::string word_s;
+      if (tokens.back() != "</s>") {
```

Comment:
similar to the <unk> thing, have a class member called eos_ (bos_ too if it's used somewhere)